A3-channel-tracing.md

```diff
@@ -0,0 +1,74 @@
+Channel Tracing
+----
+* Author(s): Noah Eisen (ncteisen@google.com)
+* Approver: markdroth
+* Status: Draft
+* Implemented in: N/A
+* Last updated: 01/02/18
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/WFDj3KeHYTI
+
+## Table of Contents
+
+  * [Abstract](#abstract)
+  * [Background](#background)
+  * [Proposal](#proposal)
+  * [Format of Exported Data](#format-of-exported-data)
+
+## Terminology
+
+This design uses the same terminology discussed in the [channelz](A14-channelz.md) design.
+
+In addition, the following terminology will be used:
+1.  an "trace event" is an interesting thing that happens to a channel. Example include things like creation, address resolution, subchannel creation, connectivity state changes.
+2. "trace data" is all of the data that is contained for one channel. This includes a list of trace events, as well as metadata like the timestamp at which the channel was created.
+3.  a "tracing object" is an in-memory data structure responsible for holding the trace data for a single channel.
+
+## Abstract
+
+The document proposes adding a dedicated, in-memory tracing object for every channel and subchannel. The tracing object will record important events in the life of a channel, like address resolution, subchannel creation, channel state changes, etc etc. The trace data from this tracing object will be made available through a new gRPC API as a JSON formatted string.
+
+Channel tracing is one aspect of the wider design, [channelz](A14-channelz.md).
+
+## Background
+
+Channel connectivity issues are the root cause of a significant portion of user reported gRPC bugs. Channel tracing will be exceedingly helpful for getting live channel data from a misbehaving program.
+
+## Proposal
+
+To implement channel tracing, implementations will use a dedicated, in-memory tracing object that is owned by each channel and subchannel. We will add an API that allows library users to retrieve all of the tracing object's data for the channel as a JSON formatted string. Implementations must expose a toggle for this behavior (for example, C++ will add a new channel argument, GRPC_ARG_CHANNEL_TRACING).
```

Comment:
For this gRFC, a specification is being described rather than an implementation being prescribed.   The difference is how programs should behave, rather than how they are written.

Here, is it really important that it is in memory?   I think the point you are trying to make is that it is ephemeral, rather than in memory.   For example, a gRPC implementation that logs channel trace events to an external database may want to not keep it in memory.

Reply:
Acknowledged and corrected.